apiserver/common/instancetypes.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/instances"
+)
+
+// InstanceTypeBackend provides the methods necessary to obtain provider
+// information.
+type InstanceTypeBackend interface {
+	environs.EnvironConfigGetter
+	Close() error
+}
+
+// modelConfigAble represents a type that can return a *config.Config.
+type modelConfigAble interface {
+	Config() (*config.Config, error)
+}
+
+// NewModelEnvironConfigGetter can create
```

Comment:
Can you please move this, modelEnvironConfigGetter, and InstanceTypeBackend to a separate file? It's only tangentially related to instance types. Also, InstanceTypeBackend should be renamed to be something to do with model/environ config.

It might be simpler just to have this as a struct of two functions? e.g.

```go
type EnvironConfigGetterFuncs struct {
    ModelConfigFunc func() (*config.Config, error)
    CloudSpecFunc func(names.ModelTag) (environs.CloudSpec, error)
}

func (f EnvironConfigGetterFuncs) ModelConfig() (*config.Config, error) {
    return f.ModelConfigFunc()
}

...
```

Just a thought, feel free to ignore that.

The EnvironConfigGetter interface is kinda weird. Both methods should probably be taking a ModelTag...